core/integration/kafka/plugin.go

```diff
@@ -0,0 +1,271 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * In applying this license CERN does not waive the privileges and
+ * immunities granted to it by virtue of its status as an
+ * Intergovernmental Organization or submit itself to any jurisdiction.
+ */
+
+//go:generate protoc --go_out=. protos/kafka.proto
+
+package kafka
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/AliceO2Group/Control/common/utils/uid"
+	"github.com/AliceO2Group/Control/core/integration"
+	kafkapb "github.com/AliceO2Group/Control/core/integration/kafka/protos"
+	"github.com/AliceO2Group/Control/core/workflow/callable"
+	"github.com/confluentinc/confluent-kafka-go/kafka"
+	"github.com/golang/protobuf/proto"
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/viper"
+	"net/url"
+	"strconv"
+	"time"
+)
+
+type Plugin struct {
+	kafkaBroker   string
+	kafkaProducer *kafka.Producer
+	envsInRunning map[string]*kafkapb.EnvInfo // env id is the key
+}
+
+func NewPlugin(endpoint string) integration.Plugin {
+	_, err := url.Parse(endpoint)
+	if err != nil {
+		log.WithField("endpoint", endpoint).
+			WithError(err).
+			Error("bad service endpoint")
+		return nil
+	}
+
+	return &Plugin{
+		kafkaBroker:   endpoint,
+		kafkaProducer: nil,
+	}
+}
+
+func (p *Plugin) GetName() string {
+	return "kafka"
+}
+
+func (p *Plugin) GetPrettyName() string {
+	return "Kafka FSM transition information service"
+}
+
+func (p *Plugin) GetEndpoint() string {
+	return viper.GetString("kafkaBroker")
+}
+
+func (p *Plugin) GetConnectionState() string {
+	if p == nil || p.kafkaProducer == nil {
+		return "UNKNOWN"
+	} else {
+		return "READY"
+	}
+}
+
+func (p *Plugin) GetData(environmentIds []uid.ID) string {
+	return ""
+}
+
+func (p *Plugin) FSMTransitionTopic(state string) string {
+	return "FSM_" + state
+}
+
+func (p *Plugin) ActiveRunsListTopic() string {
+	return "ACTIVE_RUNS"
+}
+
+func (p *Plugin) Init(instanceId string) error {
+	var err error
+	p.kafkaProducer, err = kafka.NewProducer(&kafka.ConfigMap{"bootstrap.servers": p.kafkaBroker})
+	if err != nil {
+		return fmt.Errorf("failed to initialize a kafka producer with broker '%s'. Details: %s", p.kafkaBroker, err.Error())
+	}
+	p.envsInRunning = make(map[string]*kafkapb.EnvInfo)
+	log.Info("Successfully created a kafka producer with broker '" + p.kafkaBroker + "'")
+	return nil
+}
+
+func (p *Plugin) ObjectStack(_ map[string]string) (stack map[string]interface{}) {
+	stack = make(map[string]interface{})
+	return stack
+}
+
+func parseDetectors(detectorsParam string) (detectors []string, err error) {
+	detectorsSlice := make([]string, 0)
+	bytes := []byte(detectorsParam)
+	err = json.Unmarshal(bytes, &detectorsSlice)
+	if err != nil {
+		log.WithError(err).
+			Error("error processing the detectors list")
+		return
+	}
+	return detectorsSlice, nil
+}
+
+func (p *Plugin) NewEnvStateObject(varStack map[string]string, state string) *kafkapb.EnvInfo {
```

Comment:
TBD I guess, but in principle if `state` is the actual current AliECS environment state machine state, there is probably a nicer way to get it than ask the user to call with the correct string param. See `call.go:208` and further: we preload some special values in the plugin call context, for each call, at the time of call instantiation. Right now I see no impediment to adding e.g. `__call_current_fsm_state` or similar.

Reply:
This came out to be less trivial than expected, because the code in call.go:208 is not aware of the environment state and it is not clear to me how to propagate it. One can obtain a parent role state, which is not the same (and is always UNKNOWN in this case). Thus, I took a different approach to extract the state from the `trigger` string, less clean, but still not requiring to provide `state` upfront.

I would propose to merge it as it is and once you are back, we can improve it.